applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,346 @@
+						   sel_pins[i].pin);
+		if (err) {
+			LOG_ERR("Cannot enable callback (err %d)", err);
+			irq_unlock(key);
+			return err;
+		}
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int disable_interrupts_nolock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						    sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	struct selector *selector;
+	const struct gpio_pin *sel_pins = NULL;
+	u8_t port = ARRAY_SIZE(port_map);
+
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		for (size_t j = 0; j < ARRAY_SIZE(port_map); j++) {
+			if (cb == &selectors[i].gpio_cb[j]) {
+				selector = &selectors[i];
+				sel_pins = selector->config->pins;
+				port = j;
+				break;
+			}
+		}
+	}
+	__ASSERT_NO_MSG(port < ARRAY_SIZE(port_map));
+
+	disable_interrupts_nolock(selector);
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		if ((sel_pins[i].port == port) &&
+		    (pins_mask & BIT(sel_pins[i].pin))) {
+			/* Only one pin may be set at the time. */
+			if (pins_mask & ~BIT(sel_pins[i].pin)) {
+				state = STATE_ERROR;
+			}
+			selector->state = i;
+			break;
+		}
+	}
+
+	k_delayed_work_submit(&selectors->work, 0);
+}
+
+void selector_work_fn(struct k_work *w)
+{
+	if (state == STATE_ERROR) {
+		module_set_state(MODULE_STATE_ERROR);
+		return;
+	}
+
+	struct selector *selector = CONTAINER_OF(w, struct selector, work);
+
+	selector_event_send(selector);
+
+	int err = enable_interrupts_lock(selector);
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	}
+}
+
+static int configure(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	u32_t bitmask = 0;
+
+	for (size_t i = 0; i < ARRAY_SIZE(gpio_dev); i++) {
+		for (size_t j = 0; j < selector->config->pins_size; j++) {
+			if (sel_pins[j].port == i) {
+				bitmask |= BIT(sel_pins[j].pin);
+			}
+		}
+		gpio_init_callback(&selector->gpio_cb[i], selector_isr,
+				   bitmask);
+		int err = gpio_add_callback(gpio_dev[i], &selector->gpio_cb[i]);
+
+		if (err) {
+			LOG_ERR("Cannot add callback (err %d)", err);
+			return err;
+		}
+
+		bitmask = 0;
+	}
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		int err = gpio_pin_configure(gpio_dev[sel_pins[i].port],
+					     sel_pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN |
+					     GPIO_INT | GPIO_INT_LEVEL |
+					     GPIO_INT_ACTIVE_HIGH);
+
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+	}
+
+	k_delayed_work_init(&selector->work, selector_work_fn);
+
+	return 0;
+}
+
+static int initial_read(struct selector *selector)
```

Comment:
Function should have one responsibility. Initial read suggests that the function is responsible for reading of initial state. Initialization of IRQs is unexpected side effect.

Reply:
`read_state_and_enable_interrupts` will be better?